pkg/supportbundle/supportbundle.go

```diff
@@ -0,0 +1,236 @@
+package supportbundle
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"path/filepath"
+	"runtime"
+	"runtime/pprof"
+	"strings"
+	"time"
+
+	"github.com/grafana/agent/pkg/build"
+	"github.com/grafana/agent/pkg/config"
+	"github.com/grafana/agent/pkg/server"
+	"github.com/mackerelio/go-osstat/uptime"
+	"gopkg.in/yaml.v3"
+)
+
+// Bundle collects all the data that is exposed as a support bundle.
+type Bundle struct {
+	meta                  []byte
+	config                []byte
+	agentMetrics          []byte
+	agentMetricsInstances []byte
+	agentMetricsTargets   []byte
+	agentLogsInstances    []byte
+	heapBuf               *bytes.Buffer
+	goroutineBuf          *bytes.Buffer
+	blockBuf              *bytes.Buffer
+	mutexBuf              *bytes.Buffer
+	cpuBuf                *bytes.Buffer
+}
+
+// Metadata contains general runtime information about the current Agent.
+type Metadata struct {
+	BuildVersion string                 `yaml:"build_version"`
+	OS           string                 `yaml:"os"`
+	Architecture string                 `yaml:"architecture"`
+	Uptime       float64                `yaml:"uptime"`
+	Payload      map[string]interface{} `yaml:"payload"`
+}
+
+// Export gathers the information required for the support bundle.
+func Export(ctx context.Context, enabledFeatures []string, cfg config.Config, srvAddress string, dialContext server.DialContextFunc) (*Bundle, error) {
```

Comment:
nit: just to play it safe, we should probably move the mutex here as as a global to enforce there's no way to call `Export` more than once concurrently, rather than leaving that up to the caller.

Reply:
(It might make sense to have two mutexes; one in main and the other here. Whatever works; but I think this should still be protected with a mutex somehow)